### PR TITLE
optimize: deduplication

### PR DESCRIPTION
**Pull request by 1046102779**
execute times：900000000
before optimize deduplicate: 
➜  stream git:(master) ✗ go test -v  -test.benchmem -bench=".*"
=== RUN   TestCombined
=== RUN   TestDeduplicate
=== RUN   TestReadOnce
=== RUN   TestReadUntilCanceled
--- PASS: TestReadOnce (0.00s)
--- PASS: TestDeduplicate (0.00s)
--- PASS: TestReadUntilCanceled (0.00s)
--- PASS: TestCombined (0.05s)
goos: darwin
goarch: amd64
pkg: github.com/oklog/oklog/pkg/stream
Benchmark_Deduplicate-4                1        9472183990 ns/op        3679131880 B/op 20403891 allocs/op
PASS
ok      github.com/oklog/oklog/pkg/stream       9.541s

after optimize deduplicate:  
=== RUN   TestCombined
=== RUN   TestDeduplicate
=== RUN   TestReadOnce
=== RUN   TestReadUntilCanceled
--- PASS: TestReadOnce (0.00s)
--- PASS: TestReadUntilCanceled (0.00s)
--- PASS: TestDeduplicate (0.00s)
--- PASS: TestCombined (0.05s)
goos: darwin
goarch: amd64
pkg: github.com/1046102779/oklog-1/pkg/stream
Benchmark_Deduplicate-4                1        9128090262 ns/op        3473255952 B/op 19500632 allocs/op
PASS
ok      github.com/1046102779/oklog-1/pkg/stream        9.195s

### CODE REVIEW

**Comment by 1046102779** [comment]
it maybe use  package sync.pool . haha

File: pkg/stream/deduplicate.go [hunk 1]
```diff
@@ -39,38 +40,44 @@ func Deduplicate(in <-chan []byte, window time.Duration, ticker func(time.Durati
...
 
 func (d dedupe) insert(record []byte) {
-	d.BTree.ReplaceOrInsert(item(record))
+	if len(record) <= ulid.EncodedSize {
+		panic("invalid record")
+	}
+	var node ulid.ULID
+	node.UnmarshalText(record[:ulid.EncodedSize])
+	d.BTree.ReplaceOrInsert(WrapULID{node})
+	d.ulidRecordmap[WrapULID{node}] = record
```
Comment (peterbourgon) [inline pkg/stream/deduplicate.go]: When do you delete from the ulidRecordMap? It seems like it grows without bounds!

**Comment by peterbourgon** [comment]
I'm very curious what motivated this change. Can you post some profile data, that suggests this is a bottleneck for you?

**Comment by peterbourgon** [comment]
Ping? :)

**Pull request closed without merging**